### DISCUSSION TITLE
How can we use slint with tokio in main thread event loop

### DISCUSSION THREAD

**Discussion by fluolite** [slint-ui/slint#5784] [Q&A]
Hi !
I have a question.
How can we use slint with tokio in main thread event loop!

```
fn fn_xxx() {
    let rt = tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap();
    // can not call with slint::spawn_local
    rt.spawn(async {
        // Do something in main thread, such as waiting for task result.
    });
}

fn main() -> Result<(), slint::PlatformError> {
    let app = AppWindow::new()?;
    fn_xxx();
    app.run()
}
```

**Comment by ogoffart** [comment]
I think the problem is that your runtime is destroyed before. So you need to keep the runtime alive.
And you need to use slint::spawn_local regardless because that's the main thread executor.

One way to do it is to use the [async-compat crate](https://docs.rs/async-compat/)

```rust
fn fn_xxx() {
    slint::spawn_local(async_compat::Compat::new(async {
        // Do something in main thread, such as waiting for task result.
    }));
}
```

We want to improve the [docs of spawn_local](https://docs.rs/slint/latest/slint/fn.spawn_local.html) to suggest it.

(PS: converted the question from issue to discussion, as issues are reserved for bug reports)

**Reply by fluolite** [reply]
Thanks for your reply !
Is it possible that execute a rust future in main thread with tokio's spawn but not slint::spawn_local？

**Reply by fluolite** [reply]
Such as:

````

// function in other crate has not slint dependencies
fn foo(main_rt: &tokio::runtime:Runtime, work_rt: &tokio::runtime:Runtime) {
    // can not call with slint::spawn_local
    main_rt.spawn(async {
        // work in main thread
        let result = work_rt.spawn(async{
	        // run in work thread
        }).await.unwrap();
        
        // waiting and continue in main thread
    });
}

fn main() -> Result<(), slint::PlatformError> {
    // run with main thread
    let main_rt = tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap();

    // run with work thread
    let work_rt = tokio::runtime::Builder::new_multi_thread().enable_all().build().unwrap();

    let app = AppWindow::new()?;

    foo(main_rt, work_rt);

    app.run()
}
````

**Reply by tronical** [reply]
The `spawn` call on the current-thread (`main_rt`) runtime would schedule the execution of the task, but once the Slint event loop runs, the current-thread runtime cannot actually execute the task. Therefore the current-thread runtime cannot be used in the same thread as the Slint event loop.

Do you have a specific requirement for the current-thread runtime?

**Reply by fluolite** [reply]
Thanks! 
I got it.
I'm building some generic libraries.
I hope the libraries can work with slint-ui to build a GUI App.
Some sames. it can build a command line app without slint too.
